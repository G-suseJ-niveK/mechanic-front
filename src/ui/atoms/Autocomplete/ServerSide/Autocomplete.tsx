import React, { useState, useEffect, useCallback } from 'react';
import MaterialAutocomplete from '@material-ui/lab/Autocomplete';
import { TextField, FormHelperText, FormControl, FormGroup, CircularProgress } from '@material-ui/core';
import useStyles from './Autocomplete.css';
import clsx from 'clsx';
import { Waypoint } from 'react-waypoint';

type AutocompleteProps = {
  id: string;
  label: string;
  name: string;
  items: any[];
  value: any;
  selectedValue?: any;
  defaultValue: any;
  itemText?: string;
  itemValue?: string;
  fullWidth?: boolean;
  onChange?(name: string, value: any): void;
  onBlur?(name: string, value: any): void;
  onInputChange?(value: string): Promise<boolean>;
  errors?: any;
  touched?: any;
  disabled?: boolean;
};

const Autocomplete: React.FC<AutocompleteProps> = (props: AutocompleteProps) => {
  const {
    id,
    name,
    value,
    selectedValue,
    defaultValue,
    label,
    items,
    disabled,
    onChange,
    onInputChange,
    itemText,
    itemValue,
    errors,
    touched
  } = props;
  const [currentValue, setCurrentValue] = useState<any>(null);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const classes = useStyles();

  const handleOnEnter = useCallback((args: Waypoint.CallbackArgs) => {
    if (args.previousPosition === 'below') {
    }
  }, []);

  useEffect(() => {
    let newValue = null;
    if (selectedValue === null) {
      // if itemValue exist, maybe items is a array object
      if (itemValue) {
        // find value
        newValue = items.filter((item: any) => item[itemValue] === value)[0];
      } else {
        newValue = value;
      }
    } else {
      newValue = selectedValue;
    }
    setCurrentValue(newValue || null);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const _onChange = useCallback(
    (event: any, newValue: any) => {
      setCurrentValue(newValue);
      const value = newValue !== null ? (itemValue ? newValue?.[itemValue] : String(newValue)) : defaultValue;
      onChange && onChange(name, value);
    },
    [name, onChange, itemValue, defaultValue]
  );

  const _onInputChange = useCallback(
    (event: object, value: string) => {
      if (onInputChange) {
        setIsLoading(true);
        onInputChange(value)
          .then(() => {
            setIsLoading(false);
          })
          .catch(() => {
            setIsLoading(false);
          });
      }
    },
    [onInputChange]
  );

  return (
    <FormGroup row>
      <FormControl fullWidth>
        <MaterialAutocomplete
          clearOnEscape
          autoComplete
          loading={isLoading}
          disabled={disabled}
          options={items}
          getOptionLabel={(option: any) => {
            if (option) return itemText ? option[itemText] : String(option);
            return '';
          }}
          id={id}
          value={currentValue}
          onChange={_onChange}
          onInputChange={_onInputChange}
          renderOption={(option: any) => {
            return (
              <Waypoint onEnter={handleOnEnter}>
                <div>{itemText ? option[itemText] : String(option)} </div>
              </Waypoint>
            );
          }}
          renderInput={(params: any) => (
            <TextField
              {...params}
              name={name}
              label={label}
              margin="normal"
              variant="standard"
              fullWidth
              InputProps={{
                ...params.InputProps,
                endAdornment: (
                  <>
                    {isLoading ? <CircularProgress color="inherit" size={20} /> : null}
                    {params.InputProps.endAdornment}
                  </>
                )
              }}
            />
          )}
          noOptionsText="No se encontraron coincidencias"
          loadingText="Cargando..."
          className={clsx(errors?.[name] && touched?.[name] && Boolean(errors[name]) && classes.error)}
        />
        <FormHelperText error={errors?.[name] && touched?.[name] && Boolean(errors[name])} style={{ marginTop: '0px' }}>
          {errors?.[name] && touched?.[name] ? errors[name] : ''}
        </FormHelperText>
      </FormControl>
    </FormGroup>
  );
};

export default React.memo(Autocomplete);
