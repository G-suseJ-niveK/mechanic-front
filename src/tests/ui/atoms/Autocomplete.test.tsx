import React from 'react';
import { fireEvent, render, screen, queryByAttribute } from '@testing-library/react';
import Autocomplete from '~atoms/Autocomplete/Autocomplete';

describe('Render Autocomplete', () => {
  const id: string = 'test-id';
  const name: string = 'test-name';
  const label: string = 'test-label';

  //
  // const value: any = 'aaa';

  // const defaultValue: any = 'aaa';

  const selectedValue = 'aaa';

  const items: any = ['aaa', 'bbb', 'ccc'];

  // Object
  // const jsonValue: any = {
  //   id: 1,
  //   title: 'aaa'
  // };

  // const jsonDefaultValue: any = {
  //   id: 1,
  //   title: 'aaa'
  // };

  const jsonSelectedValue = {
    id: 1,
    title: 'aaa'
  };

  const jsonItems: any = [
    {
      id: 1,
      title: 'aaa'
    },
    {
      id: 2,
      title: 'bbb'
    },
    {
      id: 3,
      title: 'ccc'
    }
  ];

  test('should render', () => {
    const { container } = render(
      <Autocomplete id={id} name={name} label={label} items={[]} value={null} defaultValue={null} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should disabled', () => {
    const { container } = render(
      <Autocomplete id={id} name={name} label={label} items={[]} value={null} defaultValue={null} disabled={true} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should render id, name and label', () => {
    const getById = queryByAttribute.bind(null, 'id');
    const getByName = queryByAttribute.bind(null, 'name');
    const { container } = render(
      <Autocomplete id={id} name={name} label={label} items={items} value={null} defaultValue={null} />
    );

    expect(getById(container, id)).toBeInstanceOf(Node);
    expect(getByName(container, name)).toBeInstanceOf(Node);
    expect(screen.getByText(label)).toBeInstanceOf(Node);
  });

  test('should render options', () => {
    render(<Autocomplete id={id} name={name} label={label} items={items} value={null} defaultValue={null} />);
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getAllByRole('option').length).toBe(3);
  });

  test('should render errors', () => {
    const errors: any = {
      'test-name': 'ERRORS'
    };

    const touched: any = {
      'test-name': true
    };

    render(
      <Autocomplete
        id={id}
        name={name}
        label={label}
        items={items}
        value={null}
        defaultValue={null}
        errors={errors}
        touched={touched}
      />
    );

    expect(screen.getByText('ERRORS')).toHaveClass('Mui-error');
    expect(screen.getByText('ERRORS')).toBeInstanceOf(Node);
  });

  test('should disableClearable', () => {
    const { container } = render(
      <Autocomplete
        id={id}
        name={name}
        label={label}
        items={items}
        value={null}
        defaultValue={null}
        disableClearable={false}
      />
    );

    expect(container).toMatchSnapshot();
  });

  describe('plain values', () => {
    test('should render selectedValue', () => {
      render(
        <Autocomplete
          id={id}
          name={name}
          label={label}
          items={items}
          value={null}
          defaultValue={null}
          selectedValue={selectedValue}
        />
      );
      expect(screen.getByRole('textbox')).toHaveValue('aaa');
    });

    test('should change value', () => {
      let currentName: string = '';
      let currentValue: any = '';

      const mockOnChange = jest.fn((testName: string, testValue: any) => {
        currentName = testName;
        currentValue = testValue;
      });

      render(
        <Autocomplete
          id={id}
          name={name}
          label={label}
          items={items}
          value={null}
          defaultValue={null}
          onChange={mockOnChange}
        />
      );

      expect(screen.getByRole('textbox')).toHaveValue('');

      fireEvent.click(screen.getByRole('button'));
      fireEvent.click(screen.getByRole('option', { name: 'bbb' }));

      expect(currentName).toBe(name);
      expect(currentValue).toBe('bbb');
      expect(screen.getByRole('textbox')).toHaveValue('bbb');
    });
  });

  describe('object values', () => {
    test('should render selectedValue', () => {
      render(
        <Autocomplete
          id={id}
          name={name}
          label={label}
          items={jsonItems}
          value={null}
          defaultValue={null}
          selectedValue={jsonSelectedValue}
        />
      );
      expect(screen.getByRole('textbox')).toHaveValue('[object Object]');
    });

    test('should render selectedValue with itemText', () => {
      render(
        <Autocomplete
          id={id}
          name={name}
          label={label}
          items={jsonItems}
          value={null}
          defaultValue={null}
          selectedValue={jsonSelectedValue}
          itemText="title"
        />
      );
      expect(screen.getByRole('textbox')).toHaveValue('aaa');
    });

    test('should change value', () => {
      render(
        <Autocomplete
          id={id}
          name={name}
          label={label}
          items={jsonItems}
          value={null}
          defaultValue={null}
          itemText="title"
        />
      );

      expect(screen.getByRole('textbox')).toHaveValue('');

      fireEvent.click(screen.getByRole('button'));
      fireEvent.click(screen.getByRole('option', { name: 'ccc' }));

      expect(screen.getByRole('textbox')).toHaveValue('ccc');
    });

    test('should change value with onChange and itemValue', () => {
      let currentName: string = '';
      let currentValue: any = '';

      const mockOnChange = jest.fn((testName: string, testValue: any) => {
        currentName = testName;
        currentValue = testValue;
      });

      render(
        <Autocomplete
          id={id}
          name={name}
          label={label}
          items={jsonItems}
          value={null}
          defaultValue={null}
          onChange={mockOnChange}
          itemText="title"
          itemValue="id"
        />
      );

      expect(screen.getByRole('textbox')).toHaveValue('');

      fireEvent.click(screen.getByRole('button'));
      fireEvent.click(screen.getByRole('option', { name: 'ccc' }));

      expect(currentName).toBe(name);
      expect(currentValue).toBe(3);
      expect(screen.getByRole('textbox')).toHaveValue('ccc');
    });
  });
});
