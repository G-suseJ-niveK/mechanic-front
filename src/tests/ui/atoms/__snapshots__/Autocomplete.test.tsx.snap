// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Autocomplete should disableClearable 1`] = `
<div>
  <div
    class="MuiFormGroup-root MuiFormGroup-row"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="test-id"
            id="test-id-label"
          >
            test-label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-autocomplete="both"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
              id="test-id"
              name="test-name"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <p
        class="MuiFormHelperText-root"
        style="margin-top: 0px;"
      />
    </div>
  </div>
</div>
`;

exports[`Render Autocomplete should disabled 1`] = `
<div>
  <div
    class="MuiFormGroup-root MuiFormGroup-row"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-disabled Mui-disabled"
            data-shrink="false"
            for="test-id"
            id="test-id-label"
          >
            test-label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-autocomplete="both"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused Mui-disabled Mui-disabled MuiInputBase-inputAdornedEnd"
              disabled=""
              id="test-id"
              name="test-name"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <p
        class="MuiFormHelperText-root"
        style="margin-top: 0px;"
      />
    </div>
  </div>
</div>
`;

exports[`Render Autocomplete should render 1`] = `
<div>
  <div
    class="MuiFormGroup-root MuiFormGroup-row"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="test-id"
            id="test-id-label"
          >
            test-label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-autocomplete="both"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
              id="test-id"
              name="test-name"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <p
        class="MuiFormHelperText-root"
        style="margin-top: 0px;"
      />
    </div>
  </div>
</div>
`;
